Sources/MQTTNIO/Connection/MQTTNewConnection.swift

```diff
@@ -0,0 +1,689 @@
+                throw MQTTError.reasonError(returnCode)
+            }
+        }
+
+        for property in connack.properties.properties {
+            // alter pingreq interval based on session expiry returned from server
+            if case .serverKeepAlive(let keepAliveInterval) = property {
+                let pingTimeout = TimeAmount.seconds(max(Int64(keepAliveInterval - 5), 5))
+                self.updatePingreqTimeout(pingTimeout)
+            }
+            // client identifier
+            if case .assignedClientIdentifier(let identifier) = property {
+                self.identifier = identifier
+            }
+            // max QoS
+            if case .maximumQoS(let qos) = property {
+                self.connectionParameters.maxQoS = qos
+            }
+            // max packet size
+            if case .maximumPacketSize(let maxPacketSize) = property {
+                self.connectionParameters.maxPacketSize = Int(maxPacketSize)
+            }
+            // supports retain
+            if case .retainAvailable(let retainValue) = property, let retainAvailable = (retainValue != 0 ? true : false) {
+                self.connectionParameters.retainAvailable = retainAvailable
+            }
+            // max topic alias
+            if case .topicAliasMaximum(let max) = property {
+                self.connectionParameters.maxTopicAlias = max
+            }
+        }
+        return connack
+    }
+
+    func processAuth(
+        _ packet: MQTTAuthPacket,
+        authWorkflow: @Sendable @escaping (MQTTAuthV5) async throws -> MQTTAuthV5
+    ) async throws -> any MQTTPacket {
+        let auth = MQTTAuthV5(reason: packet.reason, properties: packet.properties)
+        _ = try await authWorkflow(auth)
+        let responsePacket = MQTTAuthPacket(reason: packet.reason, properties: packet.properties)
+        let result = try await self.sendMessage(responsePacket) { message -> Bool in
+            guard message.type == .CONNACK || message.type == .AUTH else { throw MQTTError.failedToConnect }
+            return true
+        }
+        switch result {
+        case let connack as MQTTConnAckPacket:
+            return connack
+        case let auth as MQTTAuthPacket:
+            switch auth.reason {
+            case .continueAuthentication:
+                return try await processAuth(auth, authWorkflow: authWorkflow)
+            case .success:
+                return auth
+            default:
+                throw MQTTError.badResponse
+            }
+        default:
+            throw MQTTError.unexpectedMessage
+        }
+    }
+
+    func updatePingreqTimeout(_ timeout: TimeAmount) {
+        self.channelHandler.updatePingreqTimeout(timeout)
+    }
+
+    func sendMessage(
+        _ message: any MQTTPacket,
+        checkInbound: @escaping (MQTTPacket) throws -> Bool
+    ) async throws -> MQTTPacket {
+        try await self.channelHandler.sendMessage(message, checkInbound: checkInbound)
```

Comment:
We should have the `withCheckedThrowingContinuation` call inside of this function and remove the async version of `sendMessage` from MQTTChannelHandler. By calling the async version of sendMessage in MQTTChannelHandler you leave the actor isolation of MQTTNewConnection and are not running on the EventLoop anymore.

By doing this you can get rid of the check for `self.eventLoop.inEventLoop` in `MQTTChannelHandler.sendMessage(_:promise:checkInbound)`. Although I would replace it with a `self.eventLoop.assertInEventLoop()` to be certain we are always calling that function on the EventLoop